### PR TITLE
Added gulp task runner and extensive test suite

### PR DESCRIPTION
**Pull request by jwngr**
Thanks for maintaining this module! We use this guy in production over at [Firebase](https://www.firebase.com) and I figured I could contribute back to the project. This PR brings two main things:
  * The [`gulp` task runner](http://gulpjs.com/) to easily lint and test the repo. The default task (aka just running `gulp`) will lint the source file and then run the test suite. `gulp lint` will just run linting and, similarly, `gulp test` will just run the tests. You can also use `gulp watch` to re-run linting whenever the source file changes.
  * An extensive test suite. Your existing test suite left something to be desired and definitely didn't cover every possible edge case. I added a test suite composed of [Mocha](http://mochajs.org/), [Chai](http://chaijs.com/), and [Sinon](http://sinonjs.org/). And I covered every edge case I could think of. I made use of Sinon's [mock timers](http://sinonjs.org/docs/#clock) and [spies](http://sinonjs.org/docs/#spies), and the whole test suite runs in a cool 29 milliseconds on my machine. It even generates a code coverage report in the `/coverage/` directory which you can open up in your browser to see which lines were and were not covered.

The test suite uncovered a minor bug in the `cache.del()` method. I included a fix for it in this PR. I also made a minor change to `cache.clear()`. Namely, it now resets the `hits` and `misses` variables. This seems like desired behavior and actually makes the methods testable.

Here is the code coverage report for this PR:

```
---------------|-----------|-----------|-----------|-----------|
File           |   % Stmts |% Branches |   % Funcs |   % Lines |
---------------|-----------|-----------|-----------|-----------|
   node-cache/ |     93.33 |     82.35 |       100 |     94.52 |
      index.js |     93.33 |     82.35 |       100 |     94.52 |
---------------|-----------|-----------|-----------|-----------|
All files      |     93.33 |     82.35 |       100 |     94.52 |
---------------|-----------|-----------|-----------|-----------|
```

There are several places in the code where you have special handling for items in the cache which are expired but were not properly removed by their `setTimeout()`. Logically, I don't think these are necessary and I was unable to get any tests to ever hit this condition. Is it a race condition for this module? Is it possible we can safely remove it?

I know this a big PR, but most of it is the new test suite. I'm happy to iterate on this PR and get things to your liking. Just let me know what you like and don't like. Thanks again!

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -41,19 +41,24 @@ exports.put = function(key, value, time, timeoutCallback) {
 };
 
 exports.del = function(key) {
+  var canDelete = true;
+
   var oldRecord = cache[key];
-  var ret = true;
```
Comment (jwngr) [inline index.js]: This `ret` variable was never actually returned!

File: package.json [hunk 2]
```diff
@@ -1,28 +1,39 @@
...
+    "url": "git://github.com/ptarjan/node-cache.git"
   },
-  "scripts" : {
-    "test" : "node test.js"
+  "scripts": {
+    "test": "node test.js"
   },
-  "license" : "BSD"
+  "license": "BSD",
+  "devDependencies": {
```
Comment (jwngr) [inline package.json]: The only real diff in this file is with the `devDependencies`. `npm` caused all the whitespace changes automatically when I added the `devDependencies`.

File: test.js [hunk 3]
```diff
@@ -1,71 +1,528 @@
...
+  afterEach(function() {
+    clock.restore();
+  });
 
-var complicated = ['a',{'b':'c','d':['e',3]},'@'];
-cache.put(complicated, true);
-console.log('true == %s', cache.get(complicated));
-cache.del(complicated);
-console.log('null == %s', cache.get(complicated));
+  describe('put()', function() {
```
Comment (jwngr) [inline test.js]: You may be wondering where all the extensive tests for `put()` are. The trouble here is that we want to test all the methods in isolation, so we shouldn't use `get()`, `del()`, `size()`, etc. here since we don't know at this point in the test suite if they actually work. So we test what we can here and then in almost every other test, we put `put()`through its paces. So in the end, we do extensively test this method.

File: test.js [hunk 4]
```diff
@@ -1,71 +1,528 @@
...
+      clock.tick(1);
+      expect(cache.size()).to.equal(0);
+    });
+  });
 
-setTimeout(function() {
-  console.log('null == %s', cache.get('a'));
-  console.log('0 == %s', cache.size());
-}, 4000);
+  describe('memsize()', function() {
```
Comment (jwngr) [inline test.js]: I'm not convinced that `memsize()` is actually required. In fact, the tests for it are identical to the `size()` tests. I was unable to get them to ever get out of sync.

**Comment by ptarjan** [comment]
This all looks great. Thanks so much for your contribution. If you want your name in the contributors section, send a PR for that.
